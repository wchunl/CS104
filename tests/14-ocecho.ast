ROOT "" (0.0.0)
|   TOK_FUNCTION "" (6.22.1) function
|   |   TOK_TYPE_ID "" (6.22.1) typeid
|   |   |   TOK_VOID "void" (6.22.1) void
|   |   |   TOK_IDENT "fail" (6.22.6) variable
|   |   TOK_PARAM "(" (6.22.11) param
|   |   |   TOK_TYPE_ID "" (6.22.12) typeid
|   |   |   |   TOK_STRING "string" (6.22.12) string
|   |   |   |   TOK_IDENT "expr" (6.22.19) string variable (6.22.12)
|   |   |   TOK_TYPE_ID "" (6.22.25) typeid
|   |   |   |   TOK_STRING "string" (6.22.25) string
|   |   |   |   TOK_IDENT "file" (6.22.32) string variable (6.22.25)
|   |   |   TOK_TYPE_ID "" (6.22.38) typeid
|   |   |   |   TOK_INT "int" (6.22.38) int
|   |   |   |   TOK_IDENT "line" (6.22.42) int variable (6.22.38)
|   TOK_FUNCTION "" (6.24.1) function
|   |   TOK_TYPE_ID "" (6.24.1) typeid
|   |   |   TOK_VOID "void" (6.24.1) void
|   |   |   TOK_IDENT "putchr" (6.24.6) variable
|   |   TOK_PARAM "(" (6.24.13) param
|   |   |   TOK_TYPE_ID "" (6.24.14) typeid
|   |   |   |   TOK_INT "int" (6.24.14) int
|   |   |   |   TOK_IDENT "chr" (6.24.18) int variable (6.24.14)
|   TOK_FUNCTION "" (6.25.1) function
|   |   TOK_TYPE_ID "" (6.25.1) typeid
|   |   |   TOK_VOID "void" (6.25.1) void
|   |   |   TOK_IDENT "putint" (6.25.6) variable
|   |   TOK_PARAM "(" (6.25.13) param
|   |   |   TOK_TYPE_ID "" (6.25.14) typeid
|   |   |   |   TOK_INT "int" (6.25.14) int
|   |   |   |   TOK_IDENT "num" (6.25.18) int variable (6.25.14)
|   TOK_FUNCTION "" (6.26.1) function
|   |   TOK_TYPE_ID "" (6.26.1) typeid
|   |   |   TOK_VOID "void" (6.26.1) void
|   |   |   TOK_IDENT "putstr" (6.26.6) variable
|   |   TOK_PARAM "(" (6.26.13) param
|   |   |   TOK_TYPE_ID "" (6.26.14) typeid
|   |   |   |   TOK_STRING "string" (6.26.14) string
|   |   |   |   TOK_IDENT "str" (6.26.21) string variable (6.26.14)
|   TOK_FUNCTION "" (6.28.1) function
|   |   TOK_TYPE_ID "" (6.28.1) typeid
|   |   |   TOK_INT "int" (6.28.1) int
|   |   |   TOK_IDENT "getchr" (6.28.5) variable
|   |   TOK_PARAM "(" (6.28.11) param
|   TOK_FUNCTION "" (6.29.1) function
|   |   TOK_TYPE_ID "" (6.29.1) typeid
|   |   |   TOK_STRING "string" (6.29.1) string
|   |   |   TOK_IDENT "getstr" (6.29.8) variable
|   |   TOK_PARAM "(" (6.29.14) param
|   TOK_FUNCTION "" (6.30.1) function
|   |   TOK_TYPE_ID "" (6.30.1) typeid
|   |   |   TOK_STRING "string" (6.30.1) string
|   |   |   TOK_IDENT "getln" (6.30.8) variable
|   |   TOK_PARAM "(" (6.30.13) param
|   TOK_FUNCTION "" (7.5.1) function
|   |   TOK_TYPE_ID "" (7.5.1) typeid
|   |   |   TOK_INT "int" (7.5.1) int
|   |   |   TOK_IDENT "main" (7.5.5) variable
|   |   TOK_PARAM "(" (7.5.9) param
|   |   |   TOK_TYPE_ID "" (7.5.10) typeid
|   |   |   |   TOK_INT "int" (7.5.10) int
|   |   |   |   TOK_IDENT "argc" (7.5.14) int variable (7.5.10)
|   |   |   TOK_TYPE_ID "" (7.5.20) typeid
|   |   |   |   TOK_ARRAY "array" (7.5.20) array
|   |   |   |   |   TOK_TYPE_ID "string" (7.5.26) typeid
|   |   |   |   TOK_IDENT "argv" (7.5.34) array variable (7.5.20)
|   |   TOK_BLOCK "{" (7.5.40)
|   |   |   TOK_VARDECL "=" (7.6.13) variable lval
|   |   |   |   TOK_INT "int" (7.6.4) int
|   |   |   |   TOK_IDENT "argi" (7.6.8) int variable (7.6.13)
|   |   |   |   TOK_INTCON "1" (7.6.15) int const
|   |   |   TOK_WHILE "while" (7.7.4)
|   |   |   |   '<' "<" (7.7.16)
|   |   |   |   |   TOK_IDENT "argi" (7.7.11) int variable (7.6.13)
|   |   |   |   |   TOK_IDENT "argc" (7.7.18) int variable (7.5.10)
|   |   |   |   TOK_BLOCK "{" (7.7.24)
|   |   |   |   |   TOK_IF "if" (7.8.7)
|   |   |   |   |   |   '>' ">" (7.8.16)
|   |   |   |   |   |   |   TOK_IDENT "argi" (7.8.11) int variable (7.6.13)
|   |   |   |   |   |   |   TOK_INTCON "1" (7.8.18) int const
|   |   |   |   |   |   TOK_CALL "(" (7.8.28)
|   |   |   |   |   |   |   TOK_IDENT "putchr" (7.8.21) variable
|   |   |   |   |   |   |   TOK_CHARCON "' '" (7.8.29) int const
|   |   |   |   |   TOK_CALL "(" (7.9.14)
|   |   |   |   |   |   TOK_IDENT "putstr" (7.9.7) variable
|   |   |   |   |   |   TOK_INDEX "[" (7.9.19) lval vaddr
|   |   |   |   |   |   |   TOK_IDENT "argv" (7.9.15) array variable (7.5.20)
|   |   |   |   |   |   |   TOK_IDENT "argi" (7.9.20) int variable (7.6.13)
|   |   |   |   |   '=' "=" (7.10.12) int vreg
|   |   |   |   |   |   TOK_IDENT "argi" (7.10.7) int variable (7.6.13)
|   |   |   |   |   |   '+' "+" (7.10.19) int vreg
|   |   |   |   |   |   |   TOK_IDENT "argi" (7.10.14) int variable (7.6.13)
|   |   |   |   |   |   |   TOK_INTCON "1" (7.10.21) int const
|   |   |   TOK_CALL "(" (7.12.11)
|   |   |   |   TOK_IDENT "putstr" (7.12.4) variable
|   |   |   |   TOK_STRINGCON ""\n"" (7.12.12) string const
|   |   |   TOK_RETURN "return" (7.13.4)
|   |   |   |   TOK_INTCON "0" (7.13.11) int const
