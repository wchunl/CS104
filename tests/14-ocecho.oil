fail:     .function
          .param ptr expr
          .param ptr file
          .param int line
          .end
putchr:   .function
          .param int chr
          .end
putint:   .function
          .param int num
          .end
putstr:   .function
          .param ptr str
          .end
getchr:   .function int
          .end
getstr:   .function
          .end
getln:    .function
          .end
main:     .function int
          .param int argc
          .param ptr argv
          .local int argi
          argi = 1
.wh0:     $t0:i = argi < argc
          goto .od0 if not $t0:i
.do0:     
.if0:     $t1:i = argi > 1
          goto .el0 if not $t1:i
.th0:     putchr(' ')
.fi0:     $t2:p = argv[argi * :p]
          putstr($t2:i)
          $t3:i = argi + 1
          argi = $t3:i
.od0:     putstr("\n")
          return 0
          return 
          .end
