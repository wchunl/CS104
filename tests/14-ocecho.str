string_set[   0]:    5266873771696973564 0x1d5d018->"argi"
                    4706206694200543916 0x1d5adc8->","
                   16579916199511856337 0x1d5c488->"getln"
string_set[   1]:    2297668033614959926 0x1d5dfa8->"0"
                    3701972582333163920 0x1d5c998->"argc"
string_set[   3]:   11868120863286193964 0x1d5c6f8->"main"
                    9611428142580219387 0x1d56758->"fail"
string_set[   4]:      21606623943142546 0x1d5b238->"{"
                    4017852018407011111 0x1d5aed8->"file"
string_set[   6]:   13686721134187162920 0x1d5db98->"}"
string_set[   8]:    1370730858159036685 0x1d5d9a8->"+"
                    9386170080689500517 0x1d5ac78->"expr"
string_set[   9]:   10319632053171904911 0x1d5d228->"while"
                   14526079891305879776 0x1d5cdc8->">"
string_set[  11]:   17138325012746839705 0x1d5d248->""\n""
string_set[  12]:   15816506932648033945 0x1d5aae8->"("
string_set[  14]:    9317282556992397945 0x1d5d4a8->"if"
string_set[  17]:   18413295556186398326 0x1d5b2c8->"putint"
string_set[  18]:    7211130878751977441 0x1d5b418->"putchr"
                   12679825168387489523 0x1d5b9a8->"num"
string_set[  19]:   14643508047410943861 0x1d5ce38->"argv"
string_set[  21]:   13581035385482130682 0x1d5da68->"["
string_set[  23]:   12051031859636400188 0x1d5db78->"]"
                   11166701219533395393 0x1d5c1f8->"getstr"
                   12868445110721718657 0x1d5aef8->")"
string_set[  26]:   18172887160629263681 0x1d5d7b8->"' '"
string_set[  28]:   14635384904416252381 0x1d5de38->"return"
string_set[  29]:   12291392672316406109 0x1d5b0b8->"line"
string_set[  31]:    6142509188972423790 0x1d56738->""
                   13348923631108115385 0x1d5b738->"putstr"
string_set[  32]:   11597697714117577063 0x1d5d0c8->"="
string_set[  33]:    1583576014053361350 0x1d5bb58->"str"
string_set[  38]:   12597418120160561823 0x1d5acf8->"int"
string_set[  39]:    7296690267450279874 0x1d5b588->"chr"
string_set[  40]:   11911796782241914524 0x1d56318->"void"
string_set[  42]:    5344662657899890615 0x1d5b0d8->";"
string_set[  43]:    6814028903552918164 0x1d5b278->"array"
string_set[  44]:   13497578857207247412 0x1d5ab98->"string"
string_set[  45]:   10159970873491820195 0x1d5d178->"1"
                    1396094175042499165 0x1d5ccb8->"<"
string_set[  46]:    9201800968376702500 0x1d5bf88->"getchr"
load_factor = 0.872
bucket_count = 47
max_bucket_size = 3
