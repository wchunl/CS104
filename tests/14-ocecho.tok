#  1 "14-ocecho.oc"
#  1 "<built-in>"
#  1 "<command-line>"
#  1 "14-ocecho.oc"
#  1 "oclib.h"
# 22 "oclib.h"
   6 22.00001 261 TOK_VOID      void
   6 22.00006 282 TOK_IDENT     fail
   6 22.00011  40 '('           (
   6 22.00012 263 TOK_STRING    string
   6 22.00019 282 TOK_IDENT     expr
   6 22.00023  44 ','           ,
   6 22.00025 263 TOK_STRING    string
   6 22.00032 282 TOK_IDENT     file
   6 22.00036  44 ','           ,
   6 22.00038 262 TOK_INT       int
   6 22.00042 282 TOK_IDENT     line
   6 22.00046  41 ')'           )
   6 22.00047  59 ';'           ;
   6 24.00001 261 TOK_VOID      void
   6 24.00006 282 TOK_IDENT     putchr
   6 24.00013  40 '('           (
   6 24.00014 262 TOK_INT       int
   6 24.00018 282 TOK_IDENT     chr
   6 24.00021  41 ')'           )
   6 24.00022  59 ';'           ;
   6 25.00001 261 TOK_VOID      void
   6 25.00006 282 TOK_IDENT     putint
   6 25.00013  40 '('           (
   6 25.00014 262 TOK_INT       int
   6 25.00018 282 TOK_IDENT     num
   6 25.00021  41 ')'           )
   6 25.00022  59 ';'           ;
   6 26.00001 261 TOK_VOID      void
   6 26.00006 282 TOK_IDENT     putstr
   6 26.00013  40 '('           (
   6 26.00014 263 TOK_STRING    string
   6 26.00021 282 TOK_IDENT     str
   6 26.00024  41 ')'           )
   6 26.00025  59 ';'           ;
   6 28.00001 262 TOK_INT       int
   6 28.00005 282 TOK_IDENT     getchr
   6 28.00011  40 '('           (
   6 28.00012  41 ')'           )
   6 28.00013  59 ';'           ;
   6 29.00001 263 TOK_STRING    string
   6 29.00008 282 TOK_IDENT     getstr
   6 29.00014  40 '('           (
   6 29.00015  41 ')'           )
   6 29.00016  59 ';'           ;
   6 30.00001 263 TOK_STRING    string
   6 30.00008 282 TOK_IDENT     getln
   6 30.00013  40 '('           (
   6 30.00014  41 ')'           )
   6 30.00015  59 ';'           ;
#  4 "14-ocecho.oc"
   7 5.00001 262 TOK_INT       int
   7 5.00005 282 TOK_IDENT     main
   7 5.00009  40 '('           (
   7 5.00010 262 TOK_INT       int
   7 5.00014 282 TOK_IDENT     argc
   7 5.00018  44 ','           ,
   7 5.00020 271 TOK_ARRAY     array
   7 5.00025  60 '<'           <
   7 5.00026 263 TOK_STRING    string
   7 5.00032  62 '>'           >
   7 5.00034 282 TOK_IDENT     argv
   7 5.00038  41 ')'           )
   7 5.00040 123 '{'           {
   7 6.00004 262 TOK_INT       int
   7 6.00008 282 TOK_IDENT     argi
   7 6.00013  61 '='           =
   7 6.00015 283 TOK_INTCON    1
   7 6.00016  59 ';'           ;
   7 7.00004 267 TOK_WHILE     while
   7 7.00010  40 '('           (
   7 7.00011 282 TOK_IDENT     argi
   7 7.00016  60 '<'           <
   7 7.00018 282 TOK_IDENT     argc
   7 7.00022  41 ')'           )
   7 7.00024 123 '{'           {
   7 8.00007 265 TOK_IF        if
   7 8.00010  40 '('           (
   7 8.00011 282 TOK_IDENT     argi
   7 8.00016  62 '>'           >
   7 8.00018 283 TOK_INTCON    1
   7 8.00019  41 ')'           )
   7 8.00021 282 TOK_IDENT     putchr
   7 8.00028  40 '('           (
   7 8.00029 284 TOK_CHARCON   ' '
   7 8.00032  41 ')'           )
   7 8.00033  59 ';'           ;
   7 9.00007 282 TOK_IDENT     putstr
   7 9.00014  40 '('           (
   7 9.00015 282 TOK_IDENT     argv
   7 9.00019  91 '['           [
   7 9.00020 282 TOK_IDENT     argi
   7 9.00024  93 ']'           ]
   7 9.00025  41 ')'           )
   7 9.00026  59 ';'           ;
   7 10.00007 282 TOK_IDENT     argi
   7 10.00012  61 '='           =
   7 10.00014 282 TOK_IDENT     argi
   7 10.00019  43 '+'           +
   7 10.00021 283 TOK_INTCON    1
   7 10.00022  59 ';'           ;
   7 11.00004 125 '}'           }
   7 12.00004 282 TOK_IDENT     putstr
   7 12.00011  40 '('           (
   7 12.00012 285 TOK_STRINGCON "\n"
   7 12.00016  41 ')'           )
   7 12.00017  59 ';'           ;
   7 13.00004 268 TOK_RETURN    return
   7 13.00011 283 TOK_INTCON    0
   7 13.00012  59 ';'           ;
   7 14.00001 125 '}'           }
