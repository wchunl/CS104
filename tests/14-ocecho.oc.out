; 0.  1: # 1 "14-ocecho.oc"; 1.  0: 
; 1.  1: # 1 "<built-in>"
; 2.  1: # 1 "<command-line>"
; 3.  1: # 1 "14-ocecho.oc"
; 4.  1: 
; 4.  2: 
; 4.  3: # 1 "oclib.h" 1
; 5.  1: # 22 "oclib.h"
; 6. 22: void fail (string expr, string file, int line);
; 6. 23: 
; 6. 24: void putchr (int chr);
; 6. 25: void putint (int num);
; 6. 26: void putstr (string str);
; 6. 27: 
; 6. 28: int getchr();
; 6. 29: string getstr();
; 6. 30: string getln();
; 6. 31: # 4 "14-ocecho.oc" 2
; 7.  4: 
; 7.  5: int main(int argc, array<string> argv) {
; 7.  6:    int argi = 1;
; 7.  7:    while (argi < argc) {
; 7.  8:       if (argi > 1) putchr (' ');
; 7.  9:       putstr (argv[argi]);
; 7. 10:       argi = argi + 1;
; 7. 11:    }
; 7. 12:    putstr ("\n");
; 7. 13:    return 0;
; 7. 14: }
